_posts/2019-09-10-unbearable-lightness-of-refactoring.markdown

```diff
@@ -0,0 +1,101 @@
+---
+layout: epic
+title: "Unbearable Lightness of Refactoring"
+date: "2019-09-10"
+author: [ashkan]
+categories: [refactoring, community, culture]
+comment_id: 522
+---
+
+As engineers we are constantly in the process of building new features and improving our existing ones. Nowadays, with the help of tools and processes like code reviews one could argue the quality of the code being written has risen. At Artsy a pull request normally has one Assignee and possibly one or more Reviewers, so why we still do lot of refactoring?
+
+> There is no means of testing which decision is better, because there is no basis for comparison. We live everything as it comes, without warning, like an actor going on cold. And what can life be worth if the first rehearsal for life is life itself?
+
+― Milan Kundera, [The Unbearable Lightness of Being](https://en.wikipedia.org/wiki/The_Unbearable_Lightness_of_Being)
+
+Part of me wants to end this blogpost by Kundra’s quote, but for now lets get deeper.
+
+<!-- more -->
+
+## "The Refactor"
+Recently we've started adding SCA (Strong Customer Authentication) support to one of our services. This service is relatively young in our stack and very well reviewed. While the original code and approach looked nice and simple, as this service naturally grew and we started adding more logic to it, things got more and more complicated. During SCA support efforts, we realized its time to refactor. The code I was trying to refactor was less than a year old and originally written, well... by me!
+
+Should I be worried? embarrassed? Well, not really. As engineers when we build things we tend to look at current state of affairs, we attempt to predict the future as much as possible but the future is always changing, moving. SCA feature we were about to add to our existing logic wasn't a requirement a year ago (a year ago I didn't know what SCA was). So first thing, never be embarrassed about refactoring, because the thing you're working on is often entirely unknown and you can't expect to get something unknown totally right the first time around. Looked at in this light refactoring is healthy.
+
+Back to our SCA change, we ended up having some discussions about how to improve our already complicated logic in a way that easily supports future updates. We started by trying to separate the different steps that an order has to go through to fully get submitted and try to simplify it, weighing the pros and cons of each approach. While pretty much all of our solutions would have some cons, we managed to find our best option and started refactoring. We basically went from:
+
+```ruby
```

Comment:
It'd be nice to break this down a little bit. Isolate specific parts of the refactor and talk about why you did them.